lxd/daemon.go

```diff
@@ -2416,3 +2416,23 @@ func (d *Daemon) nodeRefreshTask(heartbeatData *cluster.APIHeartbeat, isLeader b
 
 	wg.Wait()
 }
+
+// leaderInfo returns whether this Daemon is the leader and the leader address, or an error if the leader cannot be found.
+// If the Daemon is not clustered then it is considered the leader and no address is returned.
+func (d *Daemon) leaderInfo() (bool, string, error) {
```

Comment:
In this PR I've noticed that by masking the error from d.gateway.LeaderAddress() when not clustered has required adding several explicit checks for s.ServerClustered in the code base.

I'm wondering if instead this function should return a new struct type called, say, `LeaderInfo` that contains the fields:

- Clustered bool
- Leader bool (set to true if not clustered, or clustered and is leader)
- LeaderAddress (set to empty if not clustered, or leader address if clustered)

Then we can keep all the leader/clustered related logic together.

Furthermore, when this function is stored in State.LeaderInfo, its still going to access the config from d.localConfig, so I'm wondering if we should avoid having this function on Daemon at all, and instead, store a non-exported reference to d.gateway.LeaderAddress inside State, and then have State.LeaderInfo a first-class function that accesses the existing ServerClustered, and LocalConfig.

Reply:
This all makes sense except for storing a non-exported reference to `d.gateway.LeaderAddress` inside `State`. We need to export it in order to be able to set it in the `(*Daemon).State` function.

That means we'll have both `State.LeaderAddress` and `State.LeaderInfo`, which is confusing to the caller. I'll play around with it to see if I can find something that works well.

Reply:
Why? `(*Daemon).State()` returns state.State, that doesn't have to have an exported `LeaderAddress` field, it can store that value as part of State.LeaderInfo no?

Reply:
Perhaps I'm misunderstanding what you're asking for. If I add a first class function to State to get the leader info as follows:

```go
type LeaderInfo struct {
    Clustered bool
    Leader bool
    Address string
}

func (s *State) LeaderInfo() (*LeaderInfo, error) {
	if !s.ServerClustered {
		return &state.LeaderInfo{
			Clustered: false,
			Leader:    true,
			Address:   "",
		}, nil
	}

	localClusterAddress := s.LocalConfig.ClusterAddress()
	leaderAddress, err := leaderAddressFunc()
	if err != nil {
		return nil, fmt.Errorf("Failed to get the address of the cluster leader: %w", err)
	}

	if leaderAddress == "" {
		return nil, fmt.Errorf("No leader address found")
	}

	return &state.LeaderInfo{
		Clustered: true,
		Leader:    localClusterAddress == leaderAddress,
		Address:   leaderAddress,
	}, nil
}
```

The question is where does the reference to `leaderAddressFunc` come from? I need to store it in state somewhere, but I can't set it in state via `(*Daemon).State` unless the field is exported.

I've made the changes that you've asked for except that `LeaderInfo` is not a first class function. It's being set in `(*Daemon).State`.

Reply:
Oh my bad. I did completely forget what a first class function was apparently. I've implemented it as you've asked :laughing: